### PR TITLE
Support IDP logout

### PR DESCRIPTION
**Pull request by jefff**
The PR adds support for parsing LogoutRequests at the assert endpoint, which are sent by SAML IDPs when they wish to initiate a log out. After completing the log out, a successful LogoutResponse should be sent to the relevant IDP endpoint which can be achieved by redirecting to the url provided by the new `create_logout_response_url` function.

Together, these two changes allow for full support for IDP-initiated log out.

### CODE REVIEW

File: lib/saml2.coffee [hunk 1]
```diff
@@ -81,6 +81,23 @@ create_logout_request = (issuer, name_id, session_index, destination) ->
       'samlp:SessionIndex': session_index
   .end()
 
+# Creates a LogoutResponse and returns it as a string of xml.
+create_logout_response = (issuer, in_response_to, destination, status='urn:oasis:names:tc:SAML:2.0:status:Success') ->
+  xmlbuilder.create(
+    {'samlp:LogoutResponse':
+        '@Destination': destination
```
Comment (templaedhel) [inline lib/saml2.coffee]: The indentation seems weird here.

File: lib/saml2.coffee [hunk 2]
```diff
@@ -81,6 +81,23 @@ create_logout_request = (issuer, name_id, session_index, destination) ->
       'samlp:SessionIndex': session_index
   .end()
 
+# Creates a LogoutResponse and returns it as a string of xml.
+create_logout_response = (issuer, in_response_to, destination, status='urn:oasis:names:tc:SAML:2.0:status:Success') ->
+  xmlbuilder.create(
+    {'samlp:LogoutResponse':
```
Comment (templaedhel) [inline lib/saml2.coffee]: I also don't think you need this `{`

File: lib/saml2.coffee [hunk 3]
```diff
@@ -89,13 +106,25 @@ format_pem = (key, type) ->
...
 # Takes a compressed/base64 enoded @saml_request and @private_key and signs the request using RSA-SHA256. It returns
 # the result as an object containing the query parameters.
-sign_get_request = (saml_request, private_key) ->
-  data = "SAMLRequest=" + encodeURIComponent(saml_request) + "&SigAlg=" + encodeURIComponent('http://www.w3.org/2001/04/xmldsig-more#rsa-sha256')
+sign_get_request = (saml_request, private_key, relay_state, response=false) ->
+  action = if response then "SAMLResponse" else "SAMLRequest"
+  data = "#{action}=" + encodeURIComponent(saml_request)
+  if relay_state
+    data += "&RelayState=" + encodeURIComponent(relay_state)
+  data += "&SigAlg=" + encodeURIComponent('http://www.w3.org/2001/04/xmldsig-more#rsa-sha256')
```
Comment (templaedhel) [inline lib/saml2.coffee]: Can we do this with `qs.stringify` instead of manually building it?

File: lib/saml2.coffee [hunk 4]
```diff
@@ -89,13 +106,25 @@ format_pem = (key, type) ->
...
+
+  saml_request_data = "#{action}=" + encodeURIComponent(saml_request)
+  relay_state_data = if relay_state? then "&RelayState=" + encodeURIComponent(relay_state) else ""
+  sigalg_data = "&SigAlg=" + encodeURIComponent('http://www.w3.org/2001/04/xmldsig-more#rsa-sha256')
   sign = crypto.createSign 'RSA-SHA256'
-  sign.update(data)
+  sign.update(saml_request_data + relay_state_data + sigalg_data)
```
Comment (templaedhel) [inline lib/saml2.coffee]: In fact this whole thing can be qs.stringify I think.

File: test/saml2.coffee [hunk 5]
```diff
@@ -192,6 +192,7 @@ describe 'saml2', ->
 
         expected_response =
           response_header:
+            id: '_2'
```
Comment (templaedhel) [inline test/saml2.coffee]: Are there not more tests for all these changes?

File: lib/saml2.coffee [hunk 6]
```diff
@@ -89,13 +106,25 @@ format_pem = (key, type) ->
...
 
+  saml_response_out = saml_request if response
+  saml_request_out = saml_request unless response
```
Comment (templaedhel) [inline lib/saml2.coffee]: These lines would be more clear as

```coffeescript
if response
  saml_response_out = saml_request
else
  saml_request_out = saml_request
```

File: lib/saml2.coffee [hunk 7]
```diff
@@ -313,8 +359,8 @@ module.exports.ServiceProvider =
     assert: (identity_provider, request_body, get_request..., cb) ->
       get_request =  get_request[0]
 
-      unless request_body?.SAMLResponse?
-        return setImmediate cb, new Error("Request body does not contain SAMLResponse.")
+      unless request_body?.SAMLResponse? || request_body?.SAMLRequest?
```
Comment (templaedhel) [inline lib/saml2.coffee]: `or` instead of `||`?

File: lib/saml2.coffee [hunk 8]
```diff
@@ -323,7 +369,7 @@ module.exports.ServiceProvider =
 
       async.waterfall [
         (cb_wf) ->
-          raw = new Buffer(request_body.SAMLResponse, 'base64')
+          raw = new Buffer(request_body.SAMLResponse || request_body.SAMLRequest, 'base64')
```
Comment (templaedhel) [inline lib/saml2.coffee]: `or` instead of `||`?

**Comment by templaedhel** [comment]
@jefff Can you add a PR description explaining what this PR changes? Explaining the changes in the interface, what they do, and why we need those changes? 

1. This will help me a lot as a reviewer
2. It will help other people who are curious about the library.

File: lib/saml2.coffee (hunk 3 above)
Comment (jefff) [inline lib/saml2.coffee]: No. `qs.stringify` does not have a way (that I know of) of specifying the order of the parameters. The SAML spec defines the order specifically and because the data is signed, the order MUST match the spec.

Comment (templaedhel) [inline lib/saml2.coffee]: Can we abstract this out then? A version of as.stringify that takes an array of key value pairs instead of an object for example. I don't like this manual construction littering up the code, it seems like it's easy to misunderstand or make a mistake with (forget an & for example). It also makes the code hard to parse.

File: test/saml2.coffee (hunk 5 above)
Comment (jefff) [inline test/saml2.coffee]: a1c2921

File: lib/saml2.coffee (hunk 6 above)
Comment (jefff) [inline lib/saml2.coffee]: 3e01a97

File: lib/saml2.coffee (hunk 7 above)
Comment (jefff) [inline lib/saml2.coffee]: 3e01a97

File: lib/saml2.coffee (hunk 8 above)
Comment (jefff) [inline lib/saml2.coffee]: 3e01a97

**Comment by templaedhel** [comment]
LGTM.
